src/ark.ts

```diff
@@ -1,107 +1,247 @@
+	 *
+	 * @param {string} path Bip32 Path string
+	 * @returns {Promise<string>} device compressed publicKey
+	 */
+	public async getPublicKey(path: string): Promise<string> {
+		const bip32Path = parseBip32Path(path);
+		const buffer = Buffer.alloc(1 + bip32Path.length * 4);
+		writeBip32ElementsToBuffer(bip32Path, buffer);
 
 		const response = await this.transport.send(
-			this.IDENTIFIER,
-			this.OP_GET_PUBLIC_KEY,
-			0x00,
-			this.ALG_SECP256K1,
+			ApduFlags.CLA,
+			ApduFlags.INS_GET_PUBLIC_KEY,
+			ApduFlags.P1_NON_CONFIRM,
+			ApduFlags.P2_NO_CHAINCODE,
 			buffer
 		);
 
 		return response.slice(1, 1 + response[0]).toString("hex");
 	}
 
-	public async signTransaction (path: string, hex: Buffer): Promise<string> {
-		return this.sign(path, hex, this.OP_SIGN_TRANSACTION);
+	/**
+	 * Sign a Transaction using a Ledger Device.
+	 *
+	 * @param {string} path Bip32 Path string
+	 * @param {Buffer} hex transaction payload hex
+	 * @returns {Promise<string>} payload signature
+	 */
+	public async signTransaction(path: string, hex: Buffer): Promise<string> {
+		return this.sign(path, hex, ApduFlags.INS_SIGN_TRANSACTION);
 	}
 
-	public async signMessage (path: string, hex: Buffer): Promise<string> {
-		return this.sign(path, hex, this.OP_SIGN_MESSAGE);
+	/**
+	 * Sign a Message using a Ledger Device.
+	 *
+	 * @param {string} path Bip32 Path string
+	 * @param {Buffer} hex transaction payload hex
+	 * @returns {Promise<string>} payload signature
+	 */
+	public async signMessage(path: string, hex: Buffer): Promise<string> {
+		return this.sign(path, hex, ApduFlags.INS_SIGN_MESSAGE);
 	}
 
-	private async sign (path: string, hex: Buffer, operation: number): Promise<string> {
-		if (hex.length > this.PAYLOAD_MAX) {
-			throw new Error("Payload is too large");
+	/**
+	 * Sign using a Ledger Device.
+	 *
+	 * @param {string} path Bip32 Path string
+	 * @param {Buffer} hex transaction payload hex
+	 * @param {number} instruction type of operation (e.g. Transaction, Message, etc.)
+	 * @returns {Promise<string>} payload signature
+	 * @throws {Error} if the buffer length is 0 or greater than PAYLOAD_MAX
+	 */
+	private async sign(
+		path: string,
+		hex: Buffer,
+		instruction: number
+	): Promise<string> {
+		if (hex.length === 0 || hex.length > this.PAYLOAD_MAX) {
+			throw new Error("Invalid Payload Size");
 		}
 
-		const paths = splitPath(path);
-		const hexChunks = splitToChunks(hex, (this.CHUNK_SIZE * 2) - ((2 + (paths.length * 4)) * 2));
 		const toSend = [];
-		for (const index in hexChunks) {
-			const chunk = hexChunks[index];
-			const buffer = Buffer.alloc(index === "0" ? 1 + (paths.length * 4) : 0);
 
-			if (index === "0") {
-				writePathsToBuffer(paths, buffer);
+		/** build the Apdu payload chunks */
+		const bip32Path = parseBip32Path(path);
+		const chunks = splitApduChunks(
+			hex,
+			this.CHUNK_SIZE * 2 - (2 + bip32Path.length * 4) * 2
+		);
+
+		chunks.forEach(function (chunk, index) {
```

Comment:
We tend to stick to for/in or for/of loops as those are generally faster than forEach